### PR TITLE
`reset` optional prop: `keepDirtyValues`

### PR DESCRIPTION
**Pull request by bluebill1049**
## Added: `reset` optional prop: `keepDirtyValues`

- All dirty fields will not be affected by the reset, only none interacted fields will get updated. 
- `dirtyFields` and `isDirty` form state will be remained
- Related RFC: https://github.com/react-hook-form/react-hook-form/discussions/8233
- Codesandboox: https://codesandbox.io/s/tender-panini-o8to91?file=/src/App.js

### Example:

```tsx
reset(data, {
  keepDirtyValues: true
})
```

### Todos:

- [x] RFC
- [x] tests coverage
- [x] documentation update: https://github.com/react-hook-form/documentation/pull/771

### CODE REVIEW

**Comment by Moshyfawn** [comment]
There's room for confusion in regards to the API, IMO.

Referencing the current `reset` API options naming convention, it might be seen as "keep the `dirtyFields` state as is", not "don't reset the dirty fields values"

**Comment by Moshyfawn** [comment]
On the one hand, I'd like to see a different method to reset the unmodified values only. On the other hand, tho, it wouldn't make sense to split the whole form values `reset` and the non-dirty fields one.. It's just that the `reset` API is pretty cluttered at this point

**Comment by bluebill1049** [comment]
> There's room for confusion in regards to the API, IMO.
> 
> Referencing the current `reset` API options naming convention, it might be seen as "keep the `dirtyFields` state as is", not "don't reset the dirty fields values"

It is actually the use case, which is don't reset the dirty fields (user already modified)

**Comment by bluebill1049** [comment]
> On the one hand, I'd like to see a different method to reset the unmodified values only.

I did consider that as well, but I think it potentially creates even more confusion with two APIs doing a similar thing (consider the nature of this complex use case). I don't think this is the use case that most users will encounter but if they do it's probably important to go a bit deeper on the reset API.

**Comment by Moshyfawn** [comment]
> It is actually the use case, which is don't reset the dirty fields (user already modified)

Yea, but it's not resetting the dirty fields values, not the `dirtyFields` state, right?

If you set `keepDirtyFields`, the actual behavior this feature provides is the modified **fields values** won't be changed. But all the other current `reset` options are responsible for whether or not the form's state is reset or not; like `keepDirty: true` is you keep the `isDirty` form state, `keepIsValid` is about `isValid` form's state. `reset` has `keepValues` which is clearly a flag that's responsible for input's values to be untouched. But with `keepDirtyFields` it's either the **`formState.dirtyFields` is not reset**, or **the dirty form field values**

**Comment by Moshyfawn** [comment]
> > On the one hand, I'd like to see a different method to reset the unmodified values only.
> 
> I did consider that as well, but I think it potentially creates even more confusion with two APIs doing a similar thing (consider the nature of this complex use case). I don't think this is the use case that most users will encounter but if they do it's probably important to go a bit deeper on the reset API.

Yep, fully agree

**Comment by bluebill1049** [comment]
Screen Shot 2022-04-23 at 11 17 29 pm

@Moshyfawn valid point, perhaps go with the above.

**Comment by Moshyfawn** [comment]
> Screen Shot 2022-04-23 at 11 17 29 pm
> 
> @Moshyfawn valid point, perhaps go with the above.

Oh, yeah, that is much more clear! I like it

**Comment by Moshyfawn** [comment]
@bluebill1049, what do you think about instead of having both `keepValues: boolean` and `keepDirtyValues: boolean` options, we have something like
```ts
keepValues: "all" | "unmodified" | "dirty"
```
The example `keepValues` values are provided from the top if my head. It might not be feasible to have the `"unmodified"` option, for example

**Comment by bluebill1049** [comment]
@Moshyfawn Yea i like that a lot. let's definitely change the API in v8.

**Comment by Moshyfawn** [comment]
@bluebill1049 can we generate a codesanbox build for this PR to try out different scenarios?

Are you gonna update the PR to include the `reset` options API to the one proposed in [this comment](https://github.com/react-hook-form/react-hook-form/discussions/8233#discussioncomment-2622298)? 
```ts
type ResetOptions = {
  keep: ("dirtyValues" | "isDirty" | "errors" | "values" | "defaultValues")[]
}
```

**Comment by bluebill1049** [comment]
looks like the CSB build is not working, but you can try out this one: https://codesandbox.io/s/tender-panini-o8to91?file=/src/App.js

> Are you gonna update the PR to include the reset options API to the one proposed in https://github.com/react-hook-form/react-hook-form/discussions/8233#discussioncomment-2622298?

Let's push this to v8.

**Comment by Moshyfawn** [comment]
> looks like the CSB build is not working

It might be because the PR is in "draft" state, actually

File: .eslintrc [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "@typescript-eslint/ban-ts-comment": "warn",
     "@typescript-eslint/ban-types": "off",
     "@typescript-eslint/no-explicit-any": "off",
+    "prettier/prettier": "off",
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply .eslintrc]: This has been reverted :) It was a pre-commit hook issue.

File: src/logic/createFormControl.ts [hunk 2]
```diff
@@ -1119,21 +1119,34 @@ export function createFormControl<
...
-            const fieldReference = Array.isArray(field._f.refs)
-              ? field._f.refs[0]
-              : field._f.ref;
-
-            try {
-              isHTMLElement(fieldReference) &&
-                fieldReference.closest('form')!.reset();
-              break;
-            } catch {}
+      if (keepStateOptions.keepDirtyValues) {
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/logic/createFormControl.ts:1122]: I think this is a similar scenario when you try to use `register` inside `Controller`.  If you are already keeping all the values is already a global keep for values.

[the remark this answers is not in the document]
Comment (Moshyfawn) [reply src/logic/createFormControl.ts:1122]: Fair enough!

**Review by Moshyfawn — APPROVED** [review]

**Comment by bluebill1049** [comment]
Thanks for the review @Moshyfawn this new feature driven by your RFC is awesome. ❤️

**Pull request merged**